modules/layers/src/icon-layer/icon-layer.js

```diff
@@ -121,49 +131,34 @@ export default class IconLayer extends Layer {
   updateState({oldProps, props, changeFlags}) {
     super.updateState({props, oldProps, changeFlags});
 
-    const {iconAtlas, iconMapping} = props;
+    const {iconManager} = this.state;
+    const {gl} = this.context;
+    const attributeManager = this.getAttributeManager();
 
-    if (oldProps.iconMapping !== iconMapping) {
-      const attributeManager = this.getAttributeManager();
-      attributeManager.invalidate('instanceOffsets');
-      attributeManager.invalidate('instanceIconFrames');
-      attributeManager.invalidate('instanceColorModes');
-    }
+    iconManager.updateState({oldProps, props});
 
-    if (oldProps.iconAtlas !== iconAtlas) {
-      if (iconAtlas instanceof Texture2D) {
-        iconAtlas.setParameters({
-          [GL.TEXTURE_MIN_FILTER]: DEFAULT_TEXTURE_MIN_FILTER,
-          [GL.TEXTURE_MAG_FILTER]: DEFAULT_TEXTURE_MAG_FILTER
-        });
-        this.setState({iconsTexture: iconAtlas});
-      } else if (typeof iconAtlas === 'string') {
-        loadTextures(this.context.gl, {
-          urls: [iconAtlas]
-        }).then(([texture]) => {
-          texture.setParameters({
-            [GL.TEXTURE_MIN_FILTER]: DEFAULT_TEXTURE_MIN_FILTER,
-            [GL.TEXTURE_MAG_FILTER]: DEFAULT_TEXTURE_MAG_FILTER
-          });
-          this.setState({iconsTexture: texture});
-        });
-      }
+    if (
+      changeFlags &&
+      changeFlags.updateTriggersChanged &&
+      changeFlags.updateTriggersChanged.getIcon
+    ) {
+      attributeManager.invalidate('getIcon');
```

Comment:
The `attributeManager` already handles updateTriggers itself. The key here is to tell your `iconManager` that repack is needed. There are 3 cases:
- `changeFlags.dataChanged`
- `changeFlags.updateTriggersChanged.all`
- `changeFlags.updateTriggersChanged.getIcon`